include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ * the relevant @p otLinkedBuffers and the data they contain can be reclaimed
+ * by the application. The @p otLinkedBuffers are guaranteed to be identical to
+ * those passed in to TCP via @p otTcpActiveSendByReference.
+ */
+typedef void (*otTcpSendDone)(void *aContext, otTcpActiveSocket *aSocket, otLinkedBuffer *aData);
+
+/**
+ * This callback informs that the first @p aNumBytes in the send buffer have
+ * been acknowledged by the connection peer and that their underlying memory
+ * can be reclaimed by the application.
+ *
+ * This callback is not necessary for correct TCP operation. Most applications
+ * can just rely on the @p otTcpSendDone callback. If an application wants to
+ * be able to receive fine-grained feedback as memory in the send buffer
+ * becomes available again (instead of waiting for an entire linked buffer's
+ * worth of data becomes available), it can register this callback.
+ */
+typedef void (*otTcpBytesAcked)(void *aContext, otTcpActiveSocket *aSocket, size_t aNumBytes);
+
+/**
+ * This callback informs the application that if data is added to the send
+ * buffer, some of it will be transmitted immediately without delay, as opposed
+ * to being queued for transmission at a later time. After a call to
+ * @p otTcpActiveSendByReference, an @p otTcpSendReady callback is guaranteed
+ * to be called, either immediately (if the connection is already ready) or
+ * sometime in the future (once the connection becomes ready for more data).
+ *
+ * This callback is not necessary for correct TCP operation. If more data is
+ * added to the send buffer than can be transmitted without delay, it will
+ * simply be queued for transmission at a later time. This callback should be
+ * used only in cases where some assurance is desired that data added to the
+ * send buffer will be sent soon (e.g., TCP won't wait for the recipient to
+ * ACK some other data first before sending this data out). For example, you
+ * may use this callback if you'd rather have your data be dropped than develop
+ * a backlog of data in your send buffer. But for most applications, where this
+ * isn't a concern, it's expected that one would not use this callback at all.
+ */
+typedef void (*otTcpSendReady)(void *aContext, otTcpActiveSocket *aSocket);
+
+/**
+ * This callback indicates the number of bytes available for consumption from
+ * the receive buffer. It is called whenever bytes are added to the receive
+ * buffer. If the end of the stream has been reached (i.e., if no more data
+ * will become available to read because the connection peer has closed their
+ * end of the connection for writing), then @p aEndOfStream is true. Finally,
+ * @p aBytesRemaining indicates how much capacity is left in the receive
+ * buffer to hold additional data that arrives.
+ */
+typedef void (*otTcpReceiveAvailable)(void *aContext, otTcpActiveSocket *aSocket, size_t aBytesAvailable, bool aEndOfStream, size_t aBytesRemaining);
```

Comment:
Is it possible to have `aBytesAvailable > 0` and `aEndOfStream == true` at the same time?


Does `aBytesRemaining` count the size of OOO segments? 

BTW, I can't think of a use case where `aBytesRemaining` is useful for ALP. Should we remove this information from `otTcpReceiveAvailable`?

Reply:
Yes, it's possible if you have some available for unconsumed bytes and then receive a FIN. The application needs some way to know the difference between "there are x bytes in the receive buffer and more may come" versus "there are x bytes in the receive buffer and no more bytes will be received after that."

`aBytesRemaining` counts the number of free bytes in the receive buffer, not including any OOO segments.

I added `aBytesRemaining` based on your feedback from Monday's meeting; the application also needs some feedback if the receive buffer is full and some data needs to be consumed before more data will be added to the receive buffer. Ideally, the ALP will size the receive buffer when opening the socket so that ALP doesn't need to copy data out of the receive buffer.